### DISCUSSION TITLE
[API Request] Support language tags

### DISCUSSION THREAD

**Discussion by certuna** [opensubsonic/open-subsonic-api#66] [API requests]
### Proposal description

New discussion by request @Tolriq : `language`. Working on that now in Navidrome but it makes more sense if it's in OpenSubsonic too.

Two considerations:

### Single or multi valued?
- [song](https://opensubsonic.netlify.app/docs/responses/song/): id3 and Vorbis allow multiple languages in the `TLAN/LANGUAGE` tags. But if you look at MusicBrainz Picard (and the Musicbrainz DB), language is stored single valued: a song can have only one language value, if there's multiple languages in the song it gets `mul` (= ISO code for multiple languages). Up to us to decide - is `language` a single or a multivalued property?
- [AlbumID3 [OS]](https://opensubsonic.netlify.app/docs/responses/albumid3/): what should servers do: roll up an album with 10 songs in `eng` and 1 song in `spa` as a single value `mul`? That really limits clients ability to search for "albums in English". OS doing `{eng, spa}` is better IMO.
- [ArtistID3 [OS]](https://opensubsonic.netlify.app/docs/responses/artistid3/): artist with one (or more) languages associated?
### Formatting
- it's not a free text field, id3v2.4 strictly specifies ISO 639-2, Vorbis (as usual) defines nothing, but auto-taggers like beets and Picard seem to follow id3 by only storing ISO codes
- defining this field in OpenSubsonic as ISO-code only makes it much easier for clients to localize the language name client-side, based on system language, e.g. [Intl.DisplayNames](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/DisplayNames) which also takes ISO 639-2 codes.
- [ISO 639-2](https://en.wikipedia.org/wiki/ISO_639-2) (1998, as in id3v2.4) or [ISO 639-3](https://en.wikipedia.org/wiki/ISO_639-3) (2007)? 

(edit: if we make Languages objects like Genre, then it can be up to the user, server or client to enforce ISO 639-2 validity, the API is neutral)

### Backward compatibility

- [X] No backward compatibility impact.

### API details

**one new endpoint**

`getLanguages` [OS]: `http://your-server/rest/getLanguages.view?u=demo&p=demo&v=1.13.0&c=AwesomeClientName&f=json`

```
{
  "subsonic-response": {
    "status": "ok",
    "version": "1.16.1",
    "type": "AwesomeServerName",
    "serverVersion": "0.1.3 (tag)",
    "openSubsonic": true
    "languages": {
      "language": [
        {
          "songCount": 4,
          "albumCount": 1,
          "value": "spa"
          "id": "1"
        },
        {
          "songCount": 1,
          "albumCount": 1,
          "value": "English"
          "id": "2"
        },
        {
          "songCount": 220,
          "albumCount": 2,
          "value": "eng"
          "id": "3"
        }
      ]
    }
  }
}
```

- part of the filter parameters in new endpoint `getSongs` -> `http://your-server/rest/getSongs.view?language=eng&u=demo&p=demo&v=1.13.0&c=AwesomeClientName&f=json`
- new parameter in existing `getAlbumList2`?: `http://your-server/rest/getAlbumList2.view?language=eng&type=random&u=demo&p=demo&v=1.13.0&c=AwesomeClientName&f=json`
- new parameter in existing `getArtists`?: `http://your-server/rest/getArtists.view?language=eng&u=demo&p=demo&v=1.13.0&c=AwesomeClientName&f=json`

**response:**

`language`
`languages` -> array of `language` objects

```
{
  "language": [
    {
      "songCount": 4,
      "albumCount": 1,
      "value": "spa"
      "id": "1"
    },
    {
      "songCount": 1,
      "albumCount": 1,
      "value": "English"
      "id": "2"
    },
    {
      "songCount": 220,
      "albumCount": 2,
      "value": "eng"
      "id": "3"
    }
  ]
}
```

**Comment by Tolriq** [comment]
So repeating here :)

For artists the API currently does not expose song data, if there's no genres there should not be languages either for now. If an extension is made to extend artist data with song data then it can be added but this is another discussion.

Albums should be as genre so a distinct array of the song values.

Songs tags support multiple language tags, most taggers support that, servers should not remove data because 1 website decided to. We expose the tags not MB data. If some servers wants to expose MB data they can but it must not prevent the API to return the tags.

For the format, as you said Vorbis and others do not enforce values, so enforcing something in the API brings the questions about what to do with the values that does not fit? Dropping them would be incredibly bad. Server trying to find the proper ISO code for the language, incredibly error prone.

Clients can easily still convert the ISO codes when it's 3 letters and fallback to the value if it can't.

Just the fact that ID3 says ISO 639-2 yet you propose ISO 639-3 confirm that this would not be realistic to enforce anything.

For the record I do support those fields and have already seen quite a few variations of the values that would be lost here.

**Reply by certuna** [reply]
I'm not arguing _for_ ISO 639-3, I'm raising it for discussion - I'm probably against it actually. But raising it since it's the successor of 639-2, which was the only thing that existed back in 2000 when id3v2.4 was defined. We're now in 2023 so the question is - update to 639-3 (which is largely identical anyway) or stick to 639-2 like id3. I think sticking as closely to id3v2 is best in practice, but in the end up to servers/clients devs to decide.

Enforcing certain formats in the API is already done elsewhere - for example, Vorbis also does not define track numbers and disc numbers as integers, but Subsonic does, which means servers need to enforce compliance.

Interestingly, the SubSonic API already has language codes elsewhere, for video:
```
<xs:complexType name="AudioTrack">
<xs:attribute name="id" type="xs:string" use="required"/>
<xs:attribute name="name" type="xs:string" use="optional"/>
<xs:attribute name="languageCode" type="xs:string" use="optional"/>
</xs:complexType>
```
ffmpeg for example also enforces ISO 639-2 for audio tracks in mpeg.

**Reply by Tolriq** [reply]
Well video tracks only have 1 language in most metadata format.

For track and disc the value is actually not an int since it can "track/disc" format per convention. Exposing as int makes sense as simpler for sorting and playing albums in correct order.

And the issue is not about API definition here, it's about tags and they real actual value in the real world by real users. 
We can not enforce something that would require users to retag all of their albums just to please an API.
Allowing the raw value does not prevent the clients to analyse the value and convert it to proper locale when possible.

But preventing raw value would just block tons of users and make the life of servers much more complicated if they want to try to convert those raw values to something supported in the API.

TL;DR; enforcing the value does not bring any benefits to the users or the clients.

**Reply by certuna** [reply]
But we already do - people who have tagged their FLAC files with track numbers "A1", "A2", etc have to retag to satisfy the Subsonic API, or the server performs sanitization on them before sending to Subsonic clients.

Anyway, if the stakeholders here prefer language to be a free text field in the API that's OK, it just puts the burden on the clients to filter out non-ISO 639-2 values before attempting to localize.

Also, it makes it a lot harder for a client to filter for Spanish songs, if the API sees `spa`, `Español`, `Espanol`, `Spanish` and `Spanisch` all as different languages - so IMO sanitizing the inputs server-side makes most sense from a usability pov.

**Reply by Tolriq** [reply]
You just demonstrate the opposite, this is not possible for the server to sanitize the data .. unlike track numbers.

So as I said this resume to :
- We enforce something that force users to retag all their music even if they have perfectly tailored music that works with other apps that display their tags.
- Try to do the impossible cleanup server side, with risk of data loss and hard to understand situation for users to fix.
- Allow the tag data to be exposed as the user intended to.

This is about building an API so that's it's used and is not impossible for servers to implement correctly else no one will support it or users will complains to the server that the data shown is not what they want. And asking a server to be able to match all possible variations of a language being iso codes or any possible translation in any language is not something possible nor realistic.

Exposing the tags as the user have entered them is the best solution here, because we fill the users need ...
If he tagged with ISO codes the client can easily handle the conversion in their language, if they tagged with custom values then their custom values are displayed and the user is happy. Nothing is lost here, best compromise.

We do server and clients for the users in the end, if an user have all those different Spanish tags then he probably do not care about languages at all. But if the user have taken time to properly put `Spanisch` in his tags after years of music collection then he probably care about having the data he entered.  (And as I said I support those tags and already have a few users entering full strings for the languages as most app that support them display then directly if not ISO codes)

**Reply by certuna** [reply]
> You just demonstrate the opposite, this is not possible for the server to sanitize the data .. unlike track numbers.

No that's not what I mean? The behaviour would be the same:
- User tags "A1", "A2, "B1" -> server dev has the choice to either drop the values, or write code to parse them to the integers `1`, `2`, `3` in order to fit the API specs
- User tags "Español", "Spanisch" -> server has the choice to either drop the values, or parse them to ISO 639-2 code `spa` in order to fit the API specs

That's consistent behaviour.

Of course the API could define it as a free text field, but then you'd need to parse things client-side and you lose the option to filter across different spellings of the same language.

I'd like to hear what the other server/client devs think, if there's a consensus for free text? Free text would certainly make things easier for server devs, no more API input validation needed, you can dump any string in there (well, maybe you want to be nice and sanitize for some joker taggin their files with`spa; DROP TABLE songs` before you send it to the client). But, additional work for client devs, API output validation & parsing.

**Reply by Tolriq** [reply]
There was already discussions about sanitizer for other fields in the corresponding PR.

Sanitizing numbers is easy, sanitizing arbitrary text is not. Servers can't handle all possible variations and in the end could lead to wrong data displayed to the user. If I set English in the tags I do not want a localized display in my client I chose to have it in English and not French.

Subsonic is not tied to specific online database, so relies on the tags.
If the tags allow arbitrary data and it's common for users to use them then it should be exposed. We are at the service of the user. 
If the user use language tags then he have made a choice in the value and we should respect that. If he does not and there's ton of bad values then no sanitizer will fix it.

Client can convert the iso codes. Merging all different names of the same language in the end should be left to the user not some attempt by the server/client that will never be 100%

Edit: `But, additional work for client devs, API output validation & parsing.` any client should check the data there's never a guarantee that a server will respect a random API definition. Reason all Subsonic servers already have variations. On when dealing with one server you can assume, nothing should be assumed on an opened API.

**Reply by certuna** [reply]
But you're placing additional work on client to write parsing code to unify values `English`, `eng` and `en` into the same value to be presented to the user. Or the clients don't bother parsing, but then filtering in the client for "English" will give very counterintuitive results to the user if he has 20 songs "free text" tagged `English` and 10,000 songs "properly" tagged `eng` (which is what all the auto-taggers like beets and Picard do). It will be up to the user to discover why this doesn't work.

If there's consensus for free text I'm fine with that, serverside that's easy - I can just pass through any string value in the `language` tag to the OpenSubsonic API. It's the clients that I expect may have an issue with this.

Of course, we may conclude that language should be a completely free text field in the OpenSubsonic API, and it's up to clients to validate the API output against the ISO 639-2 list and discard any non-compliant values themselves. In that case, in those clients the 20 files tagged `English` will be ignored, and the 10,000 with `eng` will remain, get mapped to their client-localized name ("English" on a US locale system, "Anglais" in on a French locale system etc), and are available for filtering on. It will be up to the user to discover why his 20 `English` tagged files don't work, and he'll retag.

As long as it's clearly documented for client devs (and users) that this is what's expected of them, then that's fine.

A second question is whether we need a `getLanguages` browing endpoint (analogous to `getGenres`) with a list of languages (and songCount/albumCount for them), so when filtering on languages, the clients can present the subset of languages that the server has, not the entire ISO 639-2 list.

Something like:
```
<xs:complexType name="Languages">
<xs:sequence>
<xs:element name="languages" type="sub:Language" minOccurs="0" maxOccurs="unbounded"/>
</xs:sequence>
</xs:complexType>
<xs:complexType name="Language" mixed="true">
<xs:attribute name="songCount" type="xs:int" use="required"/>
<xs:attribute name="albumCount" type="xs:int" use="required"/>
</xs:complexType>
```

**Reply by Tolriq** [reply]
Many people do not use auto taggers but carefully tag to control things, specially around genres. And many use some common tags as a different purpose for sorting things. Language being one of those free text multiple value field for that.

And no I'm not passing burden on the client I pass it to the user without forcing them into an arbitrary decision Subsonic have made that may not be compatible with other apps they serve the media to via other protocol.

If we expose the tags, the user have the data to fix the wrong ones. And for some it might me 10000 songs with English and 10 with eng... Why should he retag 10k songs and not fix the 10 ones he missed.

We should not enforce unnecessary things to users server or client side.

If clients have specific need they can handle it, if server drop all before client get the data then it's no more possible.

In the big picture of all needs, enforcing on server side only have disadvantages.
If the user have entered proper iso values client can already show translated values.

Edit: for `getLanguages` if there's a need for that then there should be an associated Id and languages should be returned in the language  field as an object as genres.

**Reply by certuna** [reply]
On further thinking, maybe that's a much better way to implement `language` anyway, not as a free text property of a song/album but as an object like Genre.

That way we can also sidestep this whole discussion about formatting/parsing: `English` and `eng` (ISO-mapped to "English" for US users, but to "Anglais" for French users) would show clearly as two different objects in the languages list (with different songCount/albumCount), making it clear to the user they are not the same. And then the user could fix them either way: all to hardcoded "English" spelling or to their localizable ISO code.

new endpoints:
`getLanguages`
`getSongsbyLanguage`

responses:
`language`
`languages`

so like:
```
{
  "language": [
    {
      "songCount": 1,
      "albumCount": 1,
      "value": "spa"
    },
    {
      "songCount": 4,
      "albumCount": 1,
      "value": "English"
    },
    {
      "songCount": 20,
      "albumCount": 1,
      "value": "eng"
    }
  ]
}
```

**Reply by Tolriq** [reply]
Well this work the same with a string list, but yes that's the idea.

**Reply by certuna** [reply]
let me rewrite the proposal text on top then :)

**Reply by Tolriq** [reply]
About your last edit do not forget the ids :)

And no `getSongsbyLanguage` for the moment, there's no getSongs at all yet :( We need to define the filtering stuff properly and add a single `getSongs` end point. No need to have a getSongByGenre getSongByTrackArtist getSongByAlbumArtist getSongByComposer getSongByRoleAndSubrole .... that would be insane :)

Since there's no current endpoint it's easier to directly do a proper one that we can reuse for the artists and albums after when it's matured enough.

**Reply by certuna** [reply]
You're right - I'll just leave that blank and keep it for a proper `getSongs` definition with filters?

**Reply by Tolriq** [reply]
Yes it's on the Todo, but we need to think also about improved pagination if we can in it to have a nice new base.

**Reply by certuna** [reply]
`getSongs` would be new, but for `getAlbumList2` and `getArtists` we'd just add language as an additional parameter? Or would this break legacy Subsonic servers?

**Reply by Tolriq** [reply]
for the moment no languages for getArtists as we do not yet return genre either. For Child and getAlbumList2 (And search3) we add new fields in the result that is not breaking.

For the new filters, I think the best way will be to start fresh so getAllSongs, getAllAlbums and getAllArtists with the necessary filter and pagination, so no breaking changes and tons of edge cases to handle. No name conflicts, clear definition and not adding a getAlbum99 :)

Adding new parameters is not breaking but requires the creation of an extension with definition that then becomes something we can't break. That's why it's important to try to build new strong bases. It takes times as it did for the fields and creating and pushing this project. But in the long run it's better
.

**Reply by certuna** [reply]
I'm not in a hurry, doing this right is important. But if this is the direction OpenSubsonic is going ("we do Language like Genre"), then we can already write the data structures that way server-side. Deferring to @deluan as usual since he's in the lead of Navidrome development.

**Reply by Tolriq** [reply]
Not sure many servers will support it, but as a client I already support it and I think LMS was ok to add support, so with Navidrome that's enough to validate it's wanted data. Handling it more or less as Genre (with added IDS for better handling) this support most cases and still allows the server to offer tools to the users to rename / manage they languages fields. So this makes senses.

Filtering is the second more complex step, but exposing the data to the client to at least display and for offline first client to offer all is already a good start.

**Reply by certuna** [reply]
Shouldn't be very difficult for other servers to support - I think it's mostly a chicken-and-egg problem where it doesn't make much sense to support a tag serverside when the API doesn't support it, and the lack of server support is seen as a reason not to add it to the API.

Language is a relatively straightforward property of a song, everyone knows what it means. When we get to fields like Release Type, things get much more arbitrary.

**Reply by Tolriq** [reply]
It's straightforward but triggered a ton of discussions :p

Release types are already in the OS API as `releaseTypes` and same thing we expose the tags.

**Reply by aero31aero** [reply]
I'm just chiming in for support of some implementation of this feature. It's been a pretty long discussion but I guess common principles apply:

Be restrictive in what you transmit (subsonic servers) and be lenient in what to accept (clients). Let the API spec be lenient though. It's just some text fields.

I as a user don't care whether Korean is called ko or kor or whatever else. I might as well set my music library to have three languages, A,B,C and then expect the clients to be able to filter based on lang=A, etc.
